src/core/server/http/http_server.ts

```diff
@@ -63,8 +63,12 @@ export interface HttpServerSetup {
    * (from the first registered to the last).
    */
   registerOnPostAuth: (handler: OnPostAuthHandler) => void;
-  getBasePathFor: (request: KibanaRequest | Request) => string;
-  setBasePathFor: (request: KibanaRequest | Request, basePath: string) => void;
+  basePath: {
+    get: (request: KibanaRequest | Request) => string;
+    set: (request: KibanaRequest | Request, basePath: string) => void;
+    prepend: (url: string) => string;
```

Comment:
`prepend` on client and server are different a bit. basePath can be request specific on the server side, so probably we can change the api to 
```typescript
prepend: (request: KibanaRequest | Request,  url: string) => string;
```
in our codebase https://github.com/restrry/kibana/blob/33c08f4e20b92e36c8a0df37528f410ad808d73c/x-pack/plugins/security/server/lib/authentication/providers/basic.ts#L93

Reply:
The request-specific basePath is needed because of spaces, correct? The client still needs the space prefix in its basePath, but the user can only be on a single space at a given time, where as the server may be serving multiple spaces at a time. For this reason, the `request` needs to be passed to the basePath methods on the server each time we get the basePath, but on the client, the current space is globally configured so this is not needed.

Just trying to make sure I understand.

Reply:
with handler RFC we may unified them, but atm it works as you described.

Reply:
Another way to make prepend and other helpers request specific would be to make these static functions on the BasePath class:
```
prepend: (basepath: Basepath, url) => string;
```

This way these helpers are exactly the same between client and server and there's only one type signature.

Reply:
in this case we have to create BastPath instance for every Request. don't we?

Reply:
yeah it gets a bit weird... `BasePath` as a data type isn't that useful since it's really just a string, so we could just accept a string instead:

```ts
prepend: (basepath: string, url) => string;
```

Reply:
that's the thing, end-user doesn't know `basepath` value, it is stored in the core by Spaces plugin